### PR TITLE
Add ability to parse more of "innovative" datetime formats. close #4760

### LINKED ISSUE #4760: [podcast] Unable to parse date: Mon, 14 July 2008 20:00:00 +0200
The date in some rss feeds can not be parsed.  An example is http://archiv.ulm.ccc.de/feeds/cs_mp3.xml

Reason: the month name in `<pubDate>` is not abbreviated to 3 letters as described in RFC 822 but fully written.

The follwoing fix will accept bot, an abbreviated an fully written month name:
```
diff --git a/src/core/utilities.cpp b/src/core/utilities.cpp
index 78e24e8..377d65d 100644
--- a/src/core/utilities.cpp
+++ b/src/core/utilities.cpp
@@ -580,7 +580,7 @@ QDateTime ParseRFC822DateTime(const QString& text) {
   if (ret.isValid()) return ret;
   // Because http://feeds.feedburner.com/reasonabledoubts/Msxh?format=xml
   QRegExp re2(
-      "(\\d{1,2}) ([a-zA-Z]{3}) (\\d{4}) "
+      "(\\d{1,2}) ([a-zA-Z]{3})[^ ]* (\\d{4}) "
       "(\\d{1,2}):(\\d{1,2}):(\\d{1,2})");
 
   QMap<QString, int> monthmap;
```

### PR DESCRIPTION
**Pull request by sobkas**
I couldn't stand the regexp hell, so I have removing it altogether.
Some minor style changes

### CODE REVIEW

File: src/core/utilities.cpp [hunk 1]
```diff
@@ -564,24 +559,19 @@ bool ParseUntilElement(QXmlStreamReader* reader, const QString& name) {
...
+  if (regexp.indexIn(text) == -1) {
+    return QDateTime();
   }
-  if (ret.isValid()) return ret;
-  // Because http://feeds.feedburner.com/reasonabledoubts/Msxh?format=xml
-  QRegExp re2(
-      "(\\d{1,2}) ([a-zA-Z]{3}) (\\d{4}) "
-      "(\\d{1,2}):(\\d{1,2}):(\\d{1,2})");
+
+  enum match_names {
```
Comment (davidsansome) [inline src/core/utilities.cpp]: Ugh, it's a shame QRegExp doesn't support named groups.
To make the style a little nicer can you make this an enum class, rename it MatchNames and rename the values to be in UPPER_CASE.

File: src/core/utilities.cpp [hunk 2]
```diff
@@ -564,24 +559,19 @@ bool ParseUntilElement(QXmlStreamReader* reader, const QString& name) {
...
+    days = 1,
+    months,
+    years,
+    hours,
+    minutes,
+    seconds
+  };
 
   QMap<QString, int> monthmap;
```
Comment (davidsansome) [inline src/core/utilities.cpp]: I wonder whether QMap supports initialisation list constructors?

QMap<QString, int> monthmap({
  {"Jan", 1},
  ...
});

File: src/core/utilities.cpp [hunk 3]
```diff
@@ -596,13 +586,23 @@ QDateTime ParseRFC822DateTime(const QString& text) {
...
+  monthmap["August"] = 8;
+  monthmap["September"] = 9;
+  monthmap["October"] = 10;
+  monthmap["November"] = 11;
+  monthmap["December"] = 12;
+
+  const QDate date(regexp.cap(years).toInt(), monthmap[regexp.cap(months)], regexp.cap(days).toInt());
+  const QTime time(regexp.cap(hours).toInt(), regexp.cap(minutes).toInt(), regexp.cap(seconds).toInt());
+  QDateTime date_time(date, time);
+  return date_time;
```
Comment (davidsansome) [inline src/core/utilities.cpp]: Just return QDateTime(date, time);

File: tests/utilities_test.cpp [hunk 4]
```diff
@@ -35,3 +36,12 @@ TEST(UtilitiesTest, HmacFunctions) {
   bool result_sha256 = result_hash_sha256 == QString("f7bc83f430538424b13298e6aa6fb143ef4d59a14946175997479dbc2d1a3cd8");
   EXPECT_TRUE(result_sha256);
 }
+
+TEST(UtilitiesTest, ParseRFC822DateTim) {
```
Comment (davidsansome) [inline tests/utilities_test.cpp]: Woah, a test!  I'd forgotten we had these :)

File: src/core/utilities.cpp (hunk 2 above)
Comment (sobkas) [inline src/core/utilities.cpp]: QMap from QT5.2 does support that...

**Comment by sobkas** [comment]
Made changes, hope for the best

**Pull request merged**